File: backend/src/python/adapters/data_collection.ts
/**
 * data_collection adapter — Python DataCollectionManifest → legacy
 * data_manifest the downstream parse_standardization / LLM agents
 * expect.
 */

export interface PythonCollectedDocument {
  kind: string;                     // 'financial_report' | 'activity_report' | ...
  disclosure_index: string;
  title: string;
  published_at: string;
  source_url: string;
  local_path: string;
  content_sha256: string;
  size_bytes: number;
  category?: string | null;
  subcategory?: string | null;
  summary?: string | null;
  period_label?: string | null;
  year?: number | null;
}

export interface PythonDataCollectionManifest {
  ticker: string;
  collected_at: string;
  since: string;
  until: string;
  documents?: PythonCollectedDocument[];
  sources_consulted?: Array<{ source_id?: string; url?: string; fetched_at?: string; detail?: string }>;
  errors?: string[];
  warnings?: string[];
}

export interface LegacyManifestDocument {
  kind: string;
  disclosure_id: string;
  title: string;
  period_label: string | null;
  year: number | null;
  published_at: string;
  source_url: string;
  local_path: string;
  size_bytes: number;
  content_sha256: string;
}

export interface LegacyDataCollectionOutput {
  agent_id: string;
  output_id: string;
  ticker: string;
  window: { since: string; until: string };
  data_manifest: {
    financial_reports: LegacyManifestDocument[];
    activity_reports: LegacyManifestDocument[];
    other: LegacyManifestDocument[];
  };
  document_count: number;
  sources_consulted: Array<{ source_id?: string; url?: string; detail?: string }>;
  errors: string[];
  warnings: string[];
  quality: {
    has_financials: boolean;
    has_activity: boolean;
    five_year_coverage: boolean;
  };
  review_status: string;
}


export function adaptPythonDataCollectionForLegacy(
  py: PythonDataCollectionManifest,
  outputId: string,
): LegacyDataCollectionOutput {
  const docs = py.documents ?? [];
  const mapDoc = (d: PythonCollectedDocument): LegacyManifestDocument => ({
    kind: d.kind,
    disclosure_id: d.disclosure_index,
    title: d.title,
    period_label: d.period_label ?? null,
    year: d.year ?? null,
    published_at: d.published_at,
    source_url: d.source_url,
    local_path: d.local_path,
    size_bytes: d.size_bytes,
    content_sha256: d.content_sha256,
  });

  const financial_reports = docs.filter(d => d.kind === 'financial_report').map(mapDoc);
  const activity_reports = docs.filter(d => d.kind === 'activity_report').map(mapDoc);
  const other = docs.filter(d => d.kind !== 'financial_report' && d.kind !== 'activity_report').map(mapDoc);

  const years = new Set(docs.map(d => d.year).filter((y): y is number => y !== null && y !== undefined));
  const fiveYearCoverage = years.size >= 5;

  return {
    agent_id: 'data_collection',
    output_id: outputId,
    ticker: py.ticker,
    window: { since: py.since, until: py.until },
    data_manifest: { financial_reports, activity_reports, other },
    document_count: docs.length,
    sources_consulted: py.sources_consulted ?? [],
    errors: py.errors ?? [],
    warnings: py.warnings ?? [],
    quality: {
      has_financials: financial_reports.length > 0,
      has_activity: activity_reports.length > 0,
      five_year_coverage: fiveYearCoverage,
    },
    review_status: 'pending_ceo_review',
  };
}

File: python-services/src/financex/schemas/data_collection.py
"""Schemas for data_collection — what did we fetch, from where, and where is it now.

A DataCollectionManifest is a receipt: ticker + date window + every PDF
pulled + local path + hash. Downstream (parse_standardization) reads this
to know where to dig.
"""

from __future__ import annotations

from datetime import date, datetime

from pydantic import Field

from financex.schemas.base import FinancexModel, SourceRef


class CollectedDocument(FinancexModel):
    """A single downloaded artefact."""

    # Required
    kind: str = Field(
        description="High-level kind: 'financial_report', 'activity_report', 'disclosure', 'other'.",
    )
    disclosure_index: str = Field(description="KAP disclosure index as string.")
    title: str
    published_at: datetime
    source_url: str
    local_path: str = Field(description="Absolute path on disk to the downloaded file.")
    content_sha256: str = Field(pattern=r"^[0-9a-f]{64}$")
    size_bytes: int = Field(ge=0)

    # Optional
    category: str | None = None
    subcategory: str | None = None
    summary: str | None = None
    period_label: str | None = Field(
        default=None,
        description="Reporting period this document covers, if inferable: 'FY2024', 'Q3-2024'.",
    )
    year: int | None = Field(default=None, ge=2000, le=2100)


class YearCoverageGap(FinancexModel):
    """A year for which no financial_report or activity_report was found."""

    year: int = Field(ge=2000, le=2100)
    kind: str = Field(description="'financial_report' or 'activity_report'.")
    sources_tried: list[str] = Field(
        default_factory=list,
        description="Source IDs attempted: 'kap', 'fintables', etc.",
    )
    reason: str | None = Field(
        default=None,
        description="Why the gap exists: '404_all_sources', 'rate_limited', etc.",
    )


class DataCollectionManifest(FinancexModel):
    """Summary of a data_collection run for one ticker."""

    # Required
    ticker: str
    collected_at: datetime
    since: date
    until: date
    documents: list[CollectedDocument] = Field(default_factory=list)

    # Optional diagnostics
    sources_consulted: list[SourceRef] = Field(default_factory=list)
    errors: list[str] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    coverage_gaps: list[YearCoverageGap] = Field(
        default_factory=list,
        description="Years within the window that lack a financial or activity report.",
    )

    def financial_reports(self) -> list[CollectedDocument]:
        return [d for d in self.documents if d.kind == "financial_report"]

    def activity_reports(self) -> list[CollectedDocument]:
        return [d for d in self.documents if d.kind == "activity_report"]
